### ISSUE TITLE
Question/FR/Ideas - Override default toString

### ISSUE THREAD

**Issue by andreujuanc** [ractivejs/ractive#2824]
Hello all, 

A simple binding: <p>{{ value }}</p>

Then: this.set('value', 100); Looks like it does .toString().

Currently I have done some "tricks" to get this done, stuff like defining compute values and so on.

But it would be really nice, to have a way to say to ractive engine when it takes anything that is not a string, and somehow override the method with a custom one.

`SomewhereInTheCode.stringFormatMethod = customFormatMethod || String.toString`

That way we could do something like 
`Ractive.prototype.customFormatMethod  = String.toLocaleString`

or 
```
new Ractive({
      customFormatMethod : String.toLocaleString
});
```

This way all my nonstrings bindings are formatted the way I want without having the need of surrounding everything on f( )
Something like that would be great. 

Thoughts?

**Comment by ghost** [comment]
That's exactly what methods are for. Your other options would be [computed properties](http://docs.ractivejs.org/latest/computed-properties) to represent data differently in the UI without affecting the underlying model data. Pushing further into the DOM are [decorators](http://docs.ractivejs.org/latest/decorators) which modify it on the node level.

I couldn't see the use case where one would entirely change how Ractive prints values. That seems very drastic. Can you describe a real-world use case for this?

**Comment by dagnelies** [comment]
Hi,

First, as you say, you could still write ``{{ format(...) }}``, or there is also the possibility of computed variables mentioned by @fskreuz.

It also adds complexity when you consider how it interacts with components and other plugins. Is this formatting method inherited or not? Would it be configurable? In a component with a formatting function defined, what about ``{{yield}}`` and ``{{>content}}``? Would one use the custom string function and the other not because they have different scopes?
 
As a conclusion, while it seems at first sight a meaningful enhancement, I have the feeling that the added complexity outweights the benefits. I therefore vote against 👎 .

**Comment by andreujuanc** [comment]
@fskreuz Just have lots of places where I have to manually create a computed property just for a simple toLocaleString.

@dagnelies I understand what you say but this will only make sense when type goes string.  It is true that I havent dig deep into Ractive , so I don't know much of how it parses and writes to the doom. 

> Is this formatting method inherited or not
Per instance

> Would it be configurable? 
No need, currently Ractive does .toString , just a simple way to override that method

> In a component with a formatting function defined, what about {{yield}} and {{>content}}? 
No idea, are they treated like an int, float, date? or somewhere in the code they are processed differently? Probably the second.

> Would one use the custom string function and the other not because they have different scopes?
I understand the previous question now. You mean that , what if I'm using a component that's not "mine" and it does not expects the formatting method to be a different one from toString.


My final question is: right before the data is being written to the doom, is it treated differently from each type? for example {{ someInt }} {{ someString }} and {{ yield }}, are they rendered differently? 

Can someone point me where this happens in source? If it's not much trouble .!

As always, very thankful for your replies!

**Comment by ghost** [comment]
@andreujuanc That should be in [`view/items`](https://github.com/ractivejs/ractive/tree/dev/src/view/items). A pretty good overview is found in [Views Overview](https://github.com/ractivejs/ractive/tree/dev/src/view).

TL;DR: Each construct, like an element, attribute, text, mustache - which is broken down to things like yields, partials, bindings, interpolators is an item that's managed independently. They have their own routines on how to render themselves, look up the context, how they react to data changes, how they react to DOM events. The simplest case is the [Interpolator](https://github.com/ractivejs/ractive/blob/dev/src/view/items/Interpolator.js) which just turns the data to string and pops in a text node with that string.

It appears you are doing i18n. For that use case, I think a decorator is the best way to do that. I have once done an i18n decorator for Ractive. I just passed in the data as an argument to the decorator, and the decorator managed doing the translation and appending to the element. It would look like:

```
<span as-translator="path.to.string"></span>
```

Which isn't far from how you'd normally use interpolators:

```
<span>{{ path.to.string }}</span>
```

It's also explicit, telling anyone that `path.to.string` is rendered differently from everyone else.

**Comment by dagnelies** [comment]
@andreujuanc I was indeed referring to using self-made or third-party components. It is not intuitive whether the custom format behavior should be inherited by components or not and under which conditions. 

If it's inherited by default, third-party components may break. If not, self-made components would be annoying because you redefine this custom formatting every time. And if they're nested... you get it. If it's configurable, it's complicated too. And still the behavior of `{{yield}}` vs `{{>content}}` would be unclear (this is again component specific stuff).

It also reaches more basic stuff. What about `get('value')`? Should the result be the normally formatted string or the custom formatted string? ...this is all probably the top of the iceberg.

---

Besides of this, it would be helpful to better describe what exactly you're trying to solve. As @fskreuz pointed out, there perhaps already exists a satisfying way to do it with ractive.
